Cython/Utility/ObjectHandling.c

```diff
@@ -465,7 +465,9 @@ static CYTHON_INLINE PyObject *__Pyx_GetItemInt_Fast(PyObject *o, Py_ssize_t i,
         }
     }
 #else
-    if (is_list || PySequence_Check(o)) {
+    // PySequence_GetItem behaves differently to PyObject_GetItem for i<0
+    // and possibly some other cases so can't generally be substituted
+    if (is_list || !PyMapping_Check(o)) {
         return PySequence_GetItem(o, i);
```

Comment:
Can either of `PyMapping_Check` or `PySequence_Check` really be depended on for this?

The Cython class from the issue is:
```cython
cdef class poly:

    def __len__(self):
        return 0

    cpdef long degree(self):
        return -1

    def __getitem__(self, key):
        if key < 0:
            return 0

    def leading_coefficient(self):
        return self[self.degree()]

poly().leading_coefficient()
```
This class conforms to neither of the mapping or sequence protocols. Does Cython generate `tp_as_mapping` for a type like this?

I would have expected `PyMapping_Check` to return false in this case:
https://github.com/python/cpython/blob/1198076447f35b19a9173866ccb9839f3bcf3f17/Objects/abstract.c#L2358-L2363

The change here looks as if it would still use `PySequence_GetItem` for this type unless I misunderstand it.

Reply:
Actually it looks like `PyMapping_Check` and `PySequence_Check` both return `True` in this case.

Reply:
For a Cython type it looks like `PyMapping_Check` will give true any time `__getitem__` is defined:
```cython
from cpython cimport PyMapping_Check

cdef class poly:

    def __getitem__(self, key):
        pass

    def check(self):
        return PyMapping_Check(self)
```

Reply:
When would it ever be safe to use `PySequence_GetItem` if you don't know explicitly that the object is a concrete sequence type like list or tuple?